File: src/Tree/FindDuplicateSubtrees/FindDuplicateSubtrees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    //String for No Tree
    final String NT = "X";
    //Keep track of visited subtree string key
    Set<String> visited = new HashSet<>();
    //Keep track of duplicate subtree
    Map<String, TreeNode> res = new HashMap<>();
    
    public List<TreeNode> findDuplicateSubtrees(TreeNode root) {
        postOrderTraversal(root);
        return new LinkedList<>(res.values());
    }
    
    private String postOrderTraversal(TreeNode node){
        StringBuilder sb = new StringBuilder();
        String leftSubtree = NT, rightSubtree = NT;
        
        //Traverse the left
        if(node.left != null){
            leftSubtree = postOrderTraversal(node.left);
        }
         
        //Traverse the right
        if(node.right != null){
            rightSubtree = postOrderTraversal(node.right); 
        }
        
        //Deal with current Node
        sb.append(node.val);
        
        //Build the remaining nodes in the subtree
        sb.append(",");
        sb.append(leftSubtree);
        sb.append(",");
        sb.append(rightSubtree);

        String curSubtree = sb.toString();
        saveDuplicateSubtree(node, curSubtree);
        visited.add(curSubtree);
        return curSubtree;
    }
    private void saveDuplicateSubtree(TreeNode node, String key){
        if(visited.contains(key)){
            res.put(key, node);
        }
    }
}

/**
Important lesson when deal with tree problems:
1. Remember Tree Traversal
2. Serialize tree => string can be use as a key

**/

File: src/Tree/FindDuplicateSubtrees/FindDuplicateSubtrees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findDuplicateSubtrees(self, root: Optional[TreeNode]) -> List[Optional[TreeNode]]:
        map = dict()
        unique_keys = set()
        list = []
        
        def dfs(root):
            if root is None:
                return ["X"]
            
            left_sub = dfs(root.left)
            right_sub = dfs(root.right)
            
            root_tree: List[str] = [str(root.val)]
            root_tree += left_sub + right_sub #combine 2 lists to serialize the current subtree
            
            key = ",".join(root_tree)
            
            if key in map and key not in unique_keys:
                list.append(root)
                unique_keys.add(key)
            
            map[key] = root
            
            return root_tree
        
        dfs(root)
        
        return list
